BUG: Raise TypeError when subracting DateTimeArray and other date types

KevsterAmp opened the pull request:
- [x] closes #59571 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

KevsterAmp committed:
sync to master

KevsterAmp committed:
Merge remote-tracking branch 'upstream/main'

KevsterAmp committed:
Merge remote-tracking branch 'upstream/main' into better-msg-delta-series_timestamp

KevsterAmp committed:
raise TypeError when "-" is used between DateTimeArray

KevsterAmp commented:
@rhshadrach - can you verify if my implementation here is correct? I'm not sure of this. A bit of guidance could be useful. Thanks

KevsterAmp committed:
Merge remote-tracking branch 'upstream/main' into better-msg-delta-series_timestamp

rhshadrach reviewed (changes requested):
I think this is looking good, CI failures look unrelated. They'll be fixed outside of this PR.

rhshadrach commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -18,6 +18,7 @@
 import warnings
 
 import numpy as np
+import pandas
```
We shouldn't be importing pandas at the top of a module like this - it will cause import cycles.

rhshadrach replied:
```
grep --include \*.py -nre ".*import.*DatetimeArray"
pandas/core/ops/array_ops.py:535:            from pandas.core.arrays import DatetimeArray
pandas/core/dtypes/dtypes.py:858:        from pandas.core.arrays import DatetimeArray
pandas/core/dtypes/dtypes.py:944:        from pandas.core.arrays import DatetimeArray
pandas/core/dtypes/astype.py:105:            from pandas.core.arrays import DatetimeArray
```

KevsterAmp replied:
Thanks @rhshadrach, I have removed this

KevsterAmp committed:
Merge remote-tracking branch 'upstream/main' into better-msg-delta-series_timestamp

KevsterAmp committed:
remove pandas library and improve formatting

KevsterAmp committed:
modify tests revolving this issue

KevsterAmp committed:
Merge remote-tracking branch 'upstream/main' into better-msg-delta-series_timestamp

KevsterAmp committed:
import outside top-level to handle DatetimeArray initialization

KevsterAmp committed:
replace prev error msg to new implemented error msg

KevsterAmp commented:
```python
        from pandas.core.arrays import DatetimeArray

        datetime_result = self - other
        if isinstance(datetime_result, DatetimeArray):
            raise TypeError(
                "TypeError: unsupported operand type(s) for -: "
                f"'{type(self).__name__}' and '{type(other).__name__}'"
            )
        return -(datetime_result)
```

I added an outside top-level import because I am experiencing `Unbound` errors with `Datetimearray` in this line exactly:
```python
        if isinstance(datetime_result, DatetimeArray):
```
<br>

Also, I replaced the messages with the variable type as any wildcard `.*`. Because I had trouble matching the exact dtype of the variable vs the error output. Seems like the dtype gets converted. For example, there are instances where `timedelta` variable error message contains `timedelta64` but there are cases where it returns `timedelta` as normal.

```python
            r"TypeError: unsupported operand type\(s\) for -: 'DatetimeArray' and '.*'"
```

KevsterAmp committed:
replace prev error msg on test_timedelta64 with new error msg

KevsterAmp committed:
specify dtype since test func doesn't contain multiple params/loops

rhshadrach reviewed (approved):
lgtm

KevsterAmp commented:
@mroeschke - could you review this. Thanks

mroeschke commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1495,7 +1495,15 @@ def __rsub__(self, other):
-        return -(self - other)
+        from pandas.core.arrays import DatetimeArray
+
+        datetime_result = self - other
+        if isinstance(datetime_result, DatetimeArray):
```
I'm a little nervous enforcing a particular type out of a `rsub` operation.

Could we just catch the `TypeError` and reraise it with a better message?

KevsterAmp replied:
Gotcha, I agree that it would be a better implementation

KevsterAmp committed:
catch TypeError on DateTimeArrays specifically; Improve error message

-

KevsterAmp committed:
fix tests due to changed error message

KevsterAmp commented:
My implementation raises `TypeError` on `DateTimeArrays` only. 

Since there are already other error messages for other dtypes like
```python
Cannot subtract tz-naive and tz-aware datetime-like objects.
```
If I implement it as
```python
try:
    return -(self - other)
except TypeError as e:
    raise TypeError(
        "Unsupported operand type(s) for -: "
        f"'{type(self).__name__}' and '{type(other).__name__}'"
    ) from e
```
I would have to fix all tests related to raising `tz-naive/tz-aware` datetime-like objects. Thoughts?

KevsterAmp committed:
Merge remote-tracking branch 'upstream/main' into better-msg-delta-series_timestamp

KevsterAmp commented:
Looks like my previous commit affects other datetime related operations.

Could you give me ideas on catching the error? From my understanding, when a datetime objects go into this block
```python
return -(self - other)
```
It will always raise a `TypeError`. So my previous implementation catches it before doing the operation.
```python
        # We get here with e.g. datetime objects
        from pandas.core.arrays import DatetimeArray

        datetime_result = self - other
        if isinstance(datetime_result, DatetimeArray):
            raise TypeError(
                "Unsupported operand type(s) for -: "
                f"'{type(self).__name__}' and '{type(other).__name__}'"
            )
        return -(datetime_result)
```
Though from checking the CI failures, I can't seem to trace/ I'm not familiar how my latest commit affected them. 

Guidance would be helpful. Thank you!

mroeschke commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1495,7 +1495,17 @@ def __rsub__(self, other):
+        try:
+            return -(datetime_result)
+        except TypeError as e:
+            if isinstance(datetime_result, DatetimeArray):
```
I don't think you need this `if` statement here. We'd always want to raise the exception

KevsterAmp committed:
remove if statement to always raise the exception on TypeERror

KevsterAmp committed:
Merge remote-tracking branch 'upstream/main' into better-msg-delta-series_timestamp

rhshadrach commented:
It appears this PR has gone stale. If you're interested in continuing @KevsterAmp, merge main, address comments, and the fixup the test failures above, and we'd be more than happy to reopen.
